sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiver.cs

```diff
@@ -22,8 +27,261 @@ namespace Azure.Messaging.EventHubs.Primitives
+            Argument.AssertNotNull(credential, nameof(credential));
+
+            options = options?.Clone() ?? new PartitionReceiverOptions();
+
+            OwnsConnection = true;
+            Connection = new EventHubConnection(fullyQualifiedNamespace, eventHubName, credential, options.ConnectionOptions);
+            ConsumerGroup = consumerGroup;
+            PartitionId = partitionId;
+            InitialPosition = eventPosition;
+            RetryPolicy = options.RetryOptions.ToRetryPolicy();
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PartitionReceiver"/> class.
+        /// </summary>
+        ///
+        /// <param name="consumerGroup">The name of the consumer group this client is associated with.  Events are read in the context of this group.</param>
+        /// <param name="partitionId">The identifier of the Event Hub partition from which events will be received.</param>
+        /// <param name="eventPosition">The position within the partition where the client should begin reading events.</param>
+        /// <param name="connection">The <see cref="EventHubConnection" /> connection to use for communication with the Event Hubs service.</param>
+        /// <param name="options">A set of options to apply when configuring the client.</param>
+        ///
+        public PartitionReceiver(string consumerGroup,
+                                 string partitionId,
+                                 EventPosition eventPosition,
+                                 EventHubConnection connection,
+                                 PartitionReceiverOptions options = default)
+        {
+            Argument.AssertNotNullOrEmpty(consumerGroup, nameof(consumerGroup));
+            Argument.AssertNotNullOrEmpty(partitionId, nameof(partitionId));
+            Argument.AssertNotNull(connection, nameof(connection));
+
+            options = options?.Clone() ?? new PartitionReceiverOptions();
+
+            OwnsConnection = false;
+            Connection = connection;
+            ConsumerGroup = consumerGroup;
+            PartitionId = partitionId;
+            InitialPosition = eventPosition;
+            RetryPolicy = options.RetryOptions.ToRetryPolicy();
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PartitionReceiver"/> class.
+        /// </summary>
+        ///
+        protected PartitionReceiver()
+        {
+            OwnsConnection = false;
+        }
+
+        /// <summary>
+        ///   Closes the client.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        public virtual async Task CloseAsync(CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            IsClosed = true;
```

Comment:
I'm literally just copying the behavior from other clients, but wouldn't make sense to set `IsClosed` to `true` only at the end of the method if the operation succeeded? We were doing something like this in the original `PartitionReceiver`. @jsquire

Reply:
Also - in the case where `Connection.CloseAsync` (L2664) throws an exception and the connection is not closed (say if the cancellation token is canceled) we'd report that the receiver is closed, but the connection may still be open, correct?

Reply:
Also - Should this method eagerly return in the case where `IsClosed` already returns true?  Looking at the implementation it seems like this would bottom out in `AmqpClient`, which does guard to ensure that `CloseAsync` is safe to call twice (by returning early when it was marked as closed) - but I'm wondering if we should also do an early out here.

Reply:
We went with this implementation to prevent calls to the associated operations which would potentially fail and could cause issues with corruption or indeterminate state.  For example, one thread calls `SendAsync` with a batch of messages as another calls `CloseAsync`.    We aren't likely to really understand the state of how things ended up;  this approach doesn't solve that race condition, but does help to minimizes the window where the `SendAsync` call would not fail the validation against "is the client closed."

In addition, multiple calls to `Close` which were not safe to perform and may cause issues; we never did a full exploration of the behavior of the AMQP library with concurrent calls to see how it would react.   At one point, we had been resetting `IsClosed ` to `false;` in the catch block but could not guarantee that we were in a state where it would be safe to use nor that any exceptions triggered would make sense in context.

In the worst case where there was an exception that resulted, we would potentially end up leaking a connection - which certainly isn't ideal but would time out after a period of inactivity.  Neither path is what I'd consider ideal, but I believe that the current implementation is the lesser evil.

Reply:
On the subject of eagerly returning when the flag it set... absolutely.  We should have being doing that, and I'll take a mea culpa there.   Can we please add that across clients?

Reply:
> Also - in the case where `Connection.CloseAsync` (L2664) throws an exception and the connection is not closed (say if the cancellation token is canceled) we'd report that the receiver is closed, but the connection may still be open, correct?

One quick note here, is that we only honor the cancellation token if signaled when the method was called or before we take any other actions; once the close process begins, we do not honor it, which applies to the pool of consumers as well as to the connection itself.

Reply:
Just a clarifying question about this:

> One quick note here, is that we only honor the cancellation token if signaled when the method was called or before we take any other actions; once the close process begins, we do not honor it, which applies to the pool of consumers as well as to the connection itself.

Looking at both `EventHubConnection.CloseAsync` and `AmqpClient.CloseAsync` it seems like we check for cancellation when each of these methods are called.

So I believe the case we have here is PartitionReceiver.CloseAsync calls EventHubConnection.CloseAsync which calls AmqpClient.CloseAsync which actually closes the underlying connection.

Since we check for cancellation at each one of these calls, I think it's possible that we end up not actually closing the underlying connection when it is owned, but still setting IsClosed to true, in racy cases.

Is my reading of the code correct in this case?  If so does that change how we want to structure stuff?

Reply:
I just noticed that `CancellationToken.None` wasn't being passed when closing the connection, which it should be.   You're correct in that we don't want the connection seeing our token and aborting the close because it was signaled, lest we leak.    

The same should be applied for AMQP transport constructs.   Closing cascades from the top-down, which means that should a connection be closed anything using it would likewise be force-closed, including `AmqpConsumers` and the `AmqpClient`. 

You're correct in that we've not taken steps to fully lock down all races, as we didn't want to pay the performance cost for adding synchronization across operations; in the worst case, the connection will timeout after an idle period.

Reply:
That all said, we're certainly open to ideas on how to improve the flow and lock things down better.... but I don't think that deferring the `IsClosed` would be the right trade off alone.

Reply:
I have kept the behavior of setting `IsClosed` to `true` at the start of the method like in other clients. I've removed the `cancellationToken` being passed to `Connection.CloseAsync` and I'm now returning sooner in case the client is already closed. Forgot to apply this last step across all clients, so I'll fix that.